src/variable.cpp

```diff
@@ -915,6 +937,18 @@ VariableSlice Variable::operator()(const Dim dim, const gsl::index begin,
   return {*this, dim, begin, end};
 }
 
+ConstVariableSlice Variable::reshape(const Dimensions &dims) const {
+  return {*this, dims};
+}
+
+Variable ConstVariableSlice::reshape(const Dimensions &dims) const {
+  // In general a variable slice is not contiguous. Therefore we cannot reshape
+  // without making a copy (except for special cases).
```

Comment:
What are the special cases? n x m where n == m type scenarios?

Reply:
The special case would be when the slice is contiguous, for example in a 2D array:

- Slices of thickness 1 along the outer dimension are contiguous.
- Slices of any thickness along the outer dimension are contiguous if and only if there is no slicing along the inner dimension.

Everything else is non-contiguous, consider this 2D array, S indicating the area our slice is covering:

```XXXX
XSSX
XSSX
XXXX
```

I cannot reshape the slice to shape `(4,)` since it is not representable via the normal shape/stride way of describing slices. We could in principle support this by adding another view type that provides indexed access, i.e., the view would keep an arbitrary vector of indices that are part of the slice.